arch/xtensa/src/esp32s3/esp32s3_spi_timing.c

```diff
@@ -238,129 +349,1232 @@ static void IRAM_ATTR set_flash_clock(uint8_t spi_num, uint32_t freqdiv)
+#endif
+
+#if CONFIG_ESP32S3_SPIRAM_SPEED_40M
+    *config = MSPI_TIMING_PSRAM_GET_TUNING_CONFIG(
+                ESP32S3_SPI_TIMING_CORE_CLK, 40, PSRAM_MODE);
+#elif CONFIG_ESP32S3_SPIRAM_SPEED_80M
+    *config = MSPI_TIMING_PSRAM_GET_TUNING_CONFIG(
+                ESP32S3_SPI_TIMING_CORE_CLK, 80, PSRAM_MODE);
+#elif CONFIG_ESP32S3_SPIRAM_SPEED_120M
+    *config = MSPI_TIMING_PSRAM_GET_TUNING_CONFIG(
+                ESP32S3_SPI_TIMING_CORE_CLK, 120, PSRAM_MODE);
+#endif
+
+#undef PSRAM_MODE
+}
+
+/****************************************************************************
+ * Name: config_psram_set_tuning_regs
+ *
+ * Description:
+ *   Tune PSRAM timing registers for SPI1 accessing PSRAM
+ *
+ * Input Parameters:
+ *   params - tuning timing parameters pointer
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+static void config_psram_set_tuning_regs(const tuning_param_s *params)
+{
+  /* 1. SPI_MEM_SPI_SMEM_DINx_MODE(1), SPI_MEM_SPI_SMEM_DINx_NUM(1) are
+   * meaningless SPI0 and SPI1 share the SPI_MEM_SPI_SMEM_DINx_MODE(0),
+   * SPI_MEM_SPI_SMEM_DINx_NUM(0) for PSRAM timing tuning
+   * 2. We use SPI1 to get the best PSRAM timing tuning (mode and num) config
+   */
+
+  set_psram_din_mode_num(0, params->spi_din_mode, params->spi_din_num);
+#if CONFIG_ESP32S3_SPIRAM_MODE_OCT
+
+  /* On 728, for SPI1, flash and psram share the extra dummy register */
+
+  set_flash_extra_dummy(1, params->extra_dummy_len);
+#elif CONFIG_ESP32S3_SPIRAM_MODE_QUAD
+
+  /* Update this `g_psram_extra_dummy`, the `psram_read_data` will
+   * set dummy according to this `g_psram_extra_dummy`
+   */
+
+  g_psram_extra_dummy = params->extra_dummy_len;
+
+  /* Set MSPI Quad Flash dummy */
+
+  SET_PERI_REG_MASK(SPI_MEM_USER_REG(1), SPI_MEM_USR_DUMMY);
+  SET_PERI_REG_BITS(SPI_MEM_USER1_REG(1), SPI_MEM_USR_DUMMY_CYCLELEN_V,
+                    g_psram_extra_dummy - 1, SPI_MEM_USR_DUMMY_CYCLELEN_S);
+#endif
+}
+#endif
+
+/****************************************************************************
+ * Name: sweep_for_success_sample_points
+ *
+ * Description:
+ *   Use different SPI1 timing tuning config to read data to see if current
+ *   MSPI sampling is successful.The sampling result will be stored in an
+ *   array. In this array, successful item will be 1, failed item will be 0.
+ *
+ * Input Parameters:
+ *   reference_data - reference data pointer
+ *   config         - tuning timing configuration pointer
+ *   is_flash       - is flash or not
+ *   out_array      - last success point pointer
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+static void sweep_for_success_sample_points(const uint8_t *reference_data,
+            const tuning_config_s *config, bool is_flash, uint8_t *out_array)
+{
+  uint32_t config_idx = 0;
+  uint8_t read_data[MSPI_TIMING_TEST_DATA_LEN] =
+    {
+      0
+    };
```

Comment:
```suggestion
  uint8_t read_data[MSPI_TIMING_TEST_DATA_LEN];
```
because of the line `memset(read_data, 0, MSPI_TIMING_TEST_DATA_LEN);` below

Reply:
Done